File: rock/ts-sdk/src/envs/rock_env.ts
/**
 * RockEnv - Gym-style environment interface
 */

import { envVars } from '../env_vars.js';
import { HttpUtils } from '../utils/http.js';
import { initLogger } from '../logger.js';

const logger = initLogger('rock.envs');

/**
 * Step result tuple type
 */
export type StepResult = [
  observation: unknown,
  reward: number,
  terminated: boolean,
  truncated: boolean,
  info: Record<string, unknown>
];

/**
 * Reset result tuple type
 */
export type ResetResult = [observation: unknown, info: Record<string, unknown>];

/**
 * RockEnv configuration
 */
export interface RockEnvConfig {
  envId: string;
}

/**
 * RockEnv - Gym-style environment for ROCK
 */
export class RockEnv {
  private readonly envId: string;
  private sandboxId: string | null = null;
  private isClosed = false;

  private constructor(config: RockEnvConfig) {
    this.envId = config.envId;
  }

  /**
   * Create and initialize a RockEnv instance
   *
   * @param config - Environment configuration
   * @returns Initialized RockEnv instance
   */
  static async create(config: RockEnvConfig): Promise<RockEnv> {
    const env = new RockEnv(config);
    try {
      await env.initializeEnvironment();
    } catch (e) {
      throw new Error(`Failed to initialize environment: ${e}`);
    }
    return env;
  }

  /**
   * Get the sandbox ID
   */
  getSandboxId(): string | null {
    return this.sandboxId;
  }

  /**
   * Initialize environment instance
   */
  private async initializeEnvironment(): Promise<void> {
    logger.debug(`Initializing environment: ${this.envId}`);
    const response = await HttpUtils.post<{ sandboxId: string }>(
      `${envVars.ROCK_BASE_URL}/apis/v1/envs/gem/make`,
      { 'Content-Type': 'application/json' },
      { envId: this.envId }
    );

    this.sandboxId = response.result?.sandboxId ?? null;
    if (!this.sandboxId) {
      throw new Error('Failed to get environment instance ID');
    }
  }

  /**
   * Execute an action step
   *
   * @param action - Action ID to execute
   * @returns Tuple containing observation, reward, terminated, truncated, info
   */
  async step(action: string | number): Promise<StepResult> {
    this.ensureNotClosed();

    const response = await HttpUtils.post<{
      observation: unknown;
      reward: number;
      terminated: boolean;
      truncated: boolean;
      info: Record<string, unknown>;
    }>(
      `${envVars.ROCK_BASE_URL}/apis/v1/envs/gem/step`,
      { 'Content-Type': 'application/json' },
      { sandboxId: this.sandboxId, action }
    );

    return this.parseStepResult(response.result);
  }

  /**
   * Reset environment to initial state
   *
   * @param seed - Optional random seed
   * @returns Tuple containing initial observation and info
   */
  async reset(seed?: number): Promise<ResetResult> {
    this.ensureNotClosed();

    const params: Record<string, unknown> = { sandboxId: this.sandboxId };
    if (seed !== undefined) {
      params.seed = seed;
    }

    const response = await HttpUtils.post<{
      observation: unknown;
      info: Record<string, unknown>;
    }>(
      `${envVars.ROCK_BASE_URL}/apis/v1/envs/gem/reset`,
      { 'Content-Type': 'application/json' },
      params
    );

    return this.parseResetResult(response.result);
  }

  /**
   * Close environment and clean up resources
   */
  async close(): Promise<void> {
    if (this.isClosed || !this.sandboxId) {
      return;
    }

    try {
      await HttpUtils.post(
        `${envVars.ROCK_BASE_URL}/apis/v1/envs/gem/close`,
        { 'Content-Type': 'application/json' },
        { sandboxId: this.sandboxId }
      );
    } catch (e) {
      throw new Error(`Failed to close environment: ${e}`);
    } finally {
      this.isClosed = true;
      this.sandboxId = null;
    }
  }

  /**
   * Parse step result from API response
   */
  private parseStepResult(
    data:
      | {
          observation: unknown;
          reward: number;
          terminated: boolean;
          truncated: boolean;
          info: Record<string, unknown>;
        }
      | undefined
  ): StepResult {
    if (!data) {
      throw new Error('Invalid step result: no data');
    }
    return [
      data.observation,
      data.reward,
      data.terminated,
      data.truncated,
      data.info,
    ];
  }

  /**
   * Parse reset result from API response
   */
  private parseResetResult(
    data: { observation: unknown; info: Record<string, unknown> } | undefined
  ): ResetResult {
    if (!data) {
      throw new Error('Invalid reset result: no data');
    }
    return [data.observation, data.info];
  }

  /**
   * Ensure environment is not closed
   */
  private ensureNotClosed(): void {
    if (this.isClosed) {
      throw new Error('Environment is closed');
    }
  }
}

File: rock/sdk/envs/rock_env.py
from typing import Any

import httpx

from rock import env_vars
from rock.actions import Env
from rock.logger import init_logger

logger = init_logger(__name__)


class RockEnv(Env):
    def __init__(self, env_id: str) -> None:
        """
        Initialize the environment.

        Args:
            env_id: Environment ID.

        Raises:
            Exception: Raised when environment initialization fails.
        """
        self._env_id = env_id
        self._sandbox_id: str | None = None
        self._is_closed = False
        try:
            self._initialize_environment()
        except Exception as e:
            raise Exception(f"Failed to initialize environment: {e}") from e

    def _initialize_environment(self) -> None:
        """Initialize environment instance."""
        result = self._call_admin_api("make", {"env_id": self._env_id})
        self._sandbox_id = result.get("sandbox_id")
        if not self._sandbox_id:
            raise Exception("Failed to get environment instance ID")

    def step(self, action: str) -> tuple[str, float, bool, bool, dict[str, Any]]:
        """
        Execute an action step.

        Args:
            action: Action ID to execute (0-3).

        Returns:
            Tuple containing observation, reward, termination status, truncation status, and info.

        Raises:
            Exception: Raised when action is invalid or environment is closed.
        """
        params = {
            "sandbox_id": self._sandbox_id,
            "action": action,
        }
        try:
            result = self._call_admin_api("step", params)
            return self._parse_step_result(result)
        except Exception as e:
            raise Exception(f"Failed to execute step with action {action}: {e}") from e

    def reset(self, seed: int | None = None) -> tuple[str, dict[str, Any]]:
        """
        Reset environment to initial state.

        Args:
            seed: Random seed.

        Returns:
            Tuple containing initial observation and info.

        Raises:
            Exception: Raised when reset fails.
        """
        params = {"sandbox_id": self._sandbox_id}
        if seed is not None:
            params["seed"] = seed
        try:
            result = self._call_admin_api("reset", params)
            return self._parse_reset_result(result)
        except Exception as e:
            raise Exception(f"Failed to reset environment: {e}") from e

    def close(self) -> None:
        """
        Close environment and clean up resources.

        Raises:
            Exception: Raised when closing environment fails.
        """
        if self._is_closed or not self._sandbox_id:
            return
        try:
            self._call_admin_api("close", {"sandbox_id": self._sandbox_id})
        except Exception as e:
            raise Exception(f"Failed to close environment: {e}") from e
        finally:
            self._is_closed = True
            self._sandbox_id = None

    def _parse_step_result(self, result: dict[str, Any]) -> tuple[Any, float, bool, bool, dict[str, Any]]:
        """Parse the result of step operation."""
        try:
            return (
                result["observation"],
                result["reward"],
                result["terminated"],
                result["truncated"],
                result["info"],
            )
        except (KeyError, TypeError) as e:
            raise Exception(f"Invalid step result format: {e}. Result: {result}") from e

    def _parse_reset_result(self, result: dict[str, Any]) -> tuple[Any, dict[str, Any]]:
        """Parse the result of reset operation."""
        try:
            return result["observation"], result["info"]
        except (KeyError, TypeError) as e:
            raise Exception(f"Invalid reset result format: {e}. Result: {result}") from e

    def _call_admin_api(self, endpoint: str, params: dict[str, Any]) -> dict[str, Any]:
        """
        Helper method to call Admin API.

        Args:
            endpoint: API endpoint name.
            params: Request parameters.

        Returns:
            Dictionary of API response results.

        Raises:
            Exception: Raised when API call fails.
        """
        url = f"{env_vars.ROCK_BASE_URL}/apis/v1/envs/gem/{endpoint}"
        headers = {"Content-Type": "application/json"}
        # Unified timeout configuration
        timeout = httpx.Timeout(timeout=300.0, connect=300.0, read=300.0)
        try:
            logger.debug(f"Calling Admin API {url} with params: {params}")

            with httpx.Client(timeout=timeout) as client:
                response = client.post(url, headers=headers, json=params)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPStatusError as e:
            raise Exception(f"HTTP error {e.response.status_code} for {endpoint}: {e.response.text}") from e
        except httpx.RequestError as e:
            raise Exception(f"Request error for {url}: {e}") from e
        except ValueError as e:
            raise Exception(f"JSON decode error for {url}: {e}") from e
        except Exception as e:
            raise Exception(f"Unexpected error calling {url}: {e}") from e
